stdlib/public/runtime/SwiftObject.mm

```diff
@@ -373,11 +374,62 @@ + (BOOL)conformsToProtocol:(Protocol*)proto {
 }
 
 - (NSUInteger)hash {
-  return (NSUInteger)self;
+  auto selfMetadata = _swift_getClassOfAllocated(self);
+  auto hashableConformance =
+    reinterpret_cast<const hashable_support::HashableWitnessTable *>(
+      swift_conformsToProtocolCommon(
+	selfMetadata, &hashable_support::HashableProtocolDescriptor));
+  if (hashableConformance == NULL) {
+    return (NSUInteger)self;
+  }
+  return _swift_stdlib_Hashable_hashValue_indirect(
+    &self, selfMetadata, hashableConformance);
 }
 
-- (BOOL)isEqual:(id)object {
-  return self == object;
+- (BOOL)isEqual:(id)other {
+  if (self == other) {
+    return YES;
+  }
+  // Both objects must be `SwiftObject` in ObjC
+  if (object_getClass(other) != object_getClass((id)self)) {
+    return NO;
+  }
+
+  // TODO: Bincompat check -- return NO if this is an old executable
+
+  // Get Swift type for self and other
+  auto selfMetadata = _swift_getClassOfAllocated(self);
+  auto otherMetadata = _swift_getClassOfAllocated(other);
+
+  // Find common parent Swift class
+  const ClassMetadata * parentMetadata = NULL;
+  // Collect parents of self
+  std::unordered_set<const ClassMetadata *> selfParents;
+  for (auto m = selfMetadata; m != NULL; m = m->Superclass) {
+    selfParents.emplace(m);
+  }
+  // Find first parent of other that is also a parent of self
+  for (auto m = otherMetadata; m != NULL; m = m->Superclass) {
+    if (selfParents.find(m) != selfParents.end()) {
+      parentMetadata = m;
+      break;
+    }
+  }
+  // If there's no common Swift parent class, we can't compare them
+  if (parentMetadata == NULL) {
+    return NO;
+  }
+
+  auto equatableConformance =
+    reinterpret_cast<const equatable_support::EquatableWitnessTable *>(
+      swift_conformsToProtocolCommon(
+        parentMetadata, &equatable_support::EquatableProtocolDescriptor));
+  if (equatableConformance == NULL) {
+    return NO;
+  }
+
+  return _swift_stdlib_Equatable_isEqual_indirect(
```

Comment:
This will need to be more careful, unfortunately. NSObject subclasses must always implement hashing, and `-[NSObject hash]` must always be consistent with `-[NSObject isEqual:]`. 

If the Swift type only implements `Equatable`, then `-hash` falls back to using the object's identity above, but in this implementation `isEqual:` would still forward to `T.==`. This would allow `[a isEqual: b] && a.hash != b.hash`, which would violate the NSObject protocol and wreak havoc.

So `-[SwiftObject isEqual:]` must emulate `hash`, and only forward to `==` iff the underlying type is `Hashable`; otherwise `isEqual:` must continue to compare object identities, whether or not the type is `Equatable`.

Reply:
Alternatively, could hashing default to always return 0?   That would also continue to meet the hash/equality consistency requirement.

Reply:
Apparently, NSObject's default implementation of `hashValue` simply returns `self`.  So it would make sense for that to continue to be the default for `SwiftObject` as well, in which case Karoy's right that equality can only be driven from the Hashable conformance.

Reply:
It’s a little weird since ObjC doesn’t enforce this requirement except in the sense that collections behave strangely if you mess it up. In practice, people will break this requirement knowing that they don’t use their objects in hashing collections. (And then they’ll forget about it and use a hashing collection and be sad.)

More strongly enforcing this requirement would fit with Swift’s greater emphasis on safety.

We could have SwiftObject return a 0 hash for the specific case of a type that is Equatable but not Hashable….

Reply:
Returning 0 from `-hash` would technically work! (It's not generally a good idea, though. 😅)

Reply:
Let me try that approach.   The alternative is to ignore the user-defined `==` which isn't generally a good idea either.  😄

Reply:
I've changed `-hash` to return the constant 1.

(Rationale:  people sometimes mix hashvalues of their properties via addition, and `0` has the drawback of not distinguishing types with different numbers of properties in such cases.  I don't see an argument for or against any other value, so arbitrarily chose `1`.)